Add UI Test for Orders List and Single Order Screen

thehenrybyrd opened the pull request:
Project: Critical Woo iOS UI Tests (pdcxQM-cS)

### Description
This PR adds a test for Orders, based on the structure and approach of the [Products](https://github.com/woocommerce/woocommerce-ios/pull/5674/) and [Reviews](https://github.com/woocommerce/woocommerce-ios/pull/5776) tests.
It should login using WordPress.com credentials, tap on Orders, check the Orders List, tap on the first Order, and check the Single Order that is loaded.

### Testing instructions
Pull the branch and run the test. 
Temporarily change some assertions within `OrdersScreen` or `SingleOrderScreen`, to make sure it fails when it should.
Temporarily change some data in `orders_any` to make sure it fails when it should.

Please suggest documentation for things that aren't clear!

thehenrybyrd committed:
Add readOrdersData() to MockDataReader

thehenrybyrd committed:
Add OrderDataStructs

thehenrybyrd committed:
Add verifyOrdersList() to OrdersScreen

thehenrybyrd committed:
Fix project.pbxproj

thehenrybyrd committed:
Placeholder things for testOrdersScreenLoads()

thehenrybyrd committed:
Placeholder stuff for verifyOrdersList() on OrdersScreen

thehenrybyrd committed:
Fix a few errors on OrdersScreen and MockDataReader

thehenrybyrd committed:
Fixes to testOrdersScreenLoads()

thehenrybyrd committed:
Additions to OrderDataStructs

thehenrybyrd committed:
Attempts to fix verifyOrdersList

thehenrybyrd committed:
Attempts to fix verifySingleOrder

thehenrybyrd committed:
Add an accessibilityIdentifier for single-product-cell

thehenrybyrd committed:
Comment out reference to .total

"total" isn't working because it's formatted without a ',' before the thousands.

thehenrybyrd committed:
Use NumberFormatter() to format order total

thehenrybyrd committed:
Clean up verifyOrdersList()

thehenrybyrd committed:
Make "total" a variable instead of const

since this is now manipulated by NumberFormatter()

thehenrybyrd committed:
Attempt to fix first_name reference

"Mira" appears twice on SingleOrderScreen (via Mocks), so I'm trying with assertElementExistsOnCell().

thehenrybyrd committed:
Comment out references to accessibilityIdentifier on SingleOrderScreen

thehenrybyrd committed:
Fix assertion for first name, using assertTextVisibilityCount

thehenrybyrd committed:
Cleanup extra comments

thehenrybyrd committed:
Comment out product name assertion for the moment

I'm not sure why it's failing

thehenrybyrd committed:
Change textViews to staticTexts for Product Name check

thehenrybyrd committed:
Add comments explaining the asserts

To help me remember what each assert does. May removed later.

thehenrybyrd committed:
Fix element on cell check

1. use assertElementExistsOnCell instead of verifyChildElementOnParentCell
2. use string interpolation to reference first_name + last_name

thehenrybyrd committed:
Add accessibilityIdentifier for order summary cell

thehenrybyrd committed:
Merge branch 'trunk' into add/orders-ui-tests

jostnes commented on WooCommerce/UITestsFoundation/OrderDataStructs.swift:
```diff
@@ -0,0 +1,31 @@
+public struct OrderData: Codable {
+    public let id: Int
+    public let number: String
+    public var status: String
```
Looks like this value doesn't change so can be:
```suggestion
    public let status: String
```

jostnes commented on WooCommerce/UITestsFoundation/Screens/Orders/SingleOrderScreen.swift:
```diff
@@ -8,11 +8,24 @@ public final class SingleOrderScreen: ScreenObject {
 
+    @discardableResult
+    public func verifySingleOrder(order: OrderData) throws -> Self {
+        app.assertTextVisibilityCount(textToFind: order.status, expectedCount: 1) //checks that the Mock order status appears on the screen
```
Is the comment after this line intentional? Maybe better to put on a separate line before it if needed (same comment for the few lines below as well):

```suggestion
        // checks that the Mock order status appears on the screen
        app.assertTextVisibilityCount(textToFind: order.status, expectedCount: 1) 
```

pachlava replied:
I'd vote for removing the comment, because the assertion makes it pretty clear that we expect to see the order status once.

jostnes commented on WooCommerce/UITestsFoundation/Screens/Orders/SingleOrderScreen.swift:
```diff
@@ -8,11 +8,24 @@ public final class SingleOrderScreen: ScreenObject {
+    public func verifySingleOrder(order: OrderData) throws -> Self {
+        app.assertTextVisibilityCount(textToFind: order.status, expectedCount: 1) //checks that the Mock order status appears on the screen
+        app.assertTextVisibilityCount(textToFind: order.billing.first_name, expectedCount: 2) // Check that the Mock first name ("Mira") appears twice - because this order has both billing and shipping details
+        // try app.verifyChildElementOnParentCell(parent: "single-product-cell", child: "Mira") //this shouldn't work! "single-product-cell" doesn't include the text "Mira"
```
Looks like this can be removed.

jostnes commented on WooCommerce/UITestsFoundation/XCTest+Extensions.swift:
```diff
@@ -119,10 +119,10 @@ extension XCUIElement {
-        XCTAssertTrue(count == expectedCount, "Expected '\(textToFind)' to appear \(expectedCount) times, but it appeared '\(count)' times!")
-    }
+    public func assertTextVisibilityCount(textToFind: String, expectedCount: Int = 1) {
+            let count = try! getStaticTextVisibilityCount(textToFind: textToFind)
+            XCTAssertTrue(count == expectedCount, "Expected '\(textToFind)' to appear \(expectedCount) times, but it appeared '\(count)' times!")
+        }
```
Looks to be an accidental indentation change that can be reverted.

thehenrybyrd replied:
Yep, thanks for pointing it out!

jostnes commented on WooCommerce/UITestsFoundation/XCTest+Extensions.swift:
```diff
@@ -119,10 +119,10 @@ extension XCUIElement {
-        let count = try! getStaticTextVisibilityCount(textToFind: textToFind)
-        XCTAssertTrue(count == expectedCount, "Expected '\(textToFind)' to appear \(expectedCount) times, but it appeared '\(count)' times!")
-    }
+    public func assertTextVisibilityCount(textToFind: String, expectedCount: Int = 1) {
```
👍 for adding the default value!

pachlava replied:
I'd say that using a default value in this particular case does not improve the explicitness of code. A person that sees this the first time:

```swift
app.assertTextVisibilityCount(textToFind: String(orders[0].id))
```

Will most likely want to take a look into the function to make sure that expected count is `1` for the default case. I'd suggest either adding an extra method (like `assertTextShownOnce` or a better name) which will call `assertTextVisibilityCount` in its turn, and use the new method for the cases when the expected number is one. Or, always provide the expected count value.

Another suggestion is that you could utilize a possibility to omit parameter names, and also have the `external` and `internal` parameter names, and adjust the naming a bit (because the name like `assertTextVisibilityCount ` already implies that one of the parameter will be text, and the other will be count):

```suggestion
    public func assertVisibilityCount(forText text: String, _ expectedCount: Int = 1) {
```

Which will allow you to have a call looking like this:

`app.assertVisibilityCount(forText: order.status, 1)`

(naming are just the examples)

pachlava replied:
Just realized that this is an old method... So my suggestions from above are completely optional.

thehenrybyrd replied:
I agree that adding the default value doesn't improve the explicitness/readability, however I needed a way to use `expectedCount: 2` since I'm calling this in SingleOrderScreen in order to check the first name, which appears twice in this case. Since most of the rest of the calls to this function do expect only 1 count, I made that the default.
Then, in SingleOrderScreen, I could assert:
`app.assertTextVisibilityCount(textToFind: order.billing.first_name, expectedCount: 2)`

jostnes commented on WooCommerce/WooCommerceUITests/Flows/MockDataReader.swift:
```diff
@@ -51,4 +51,24 @@ class GetMocks {
+        for index in 0..<updatedData.count {
+                    var total = updatedData[index].total
+                    print("total is: \(total)")
+                    print("total after converted: \(Double(total) ?? 0)")
```
The print commands can be removed.

jostnes commented on WooCommerce/WooCommerceUITests/Tests/OrdersTests.swift:
```diff
@@ -1,5 +1,6 @@
 import UITestsFoundation
 import XCTest
+import Networking
```
This can probably also be removed. Could be an accidental add?

thehenrybyrd replied:
Yep, no idea how that got there 😆

jostnes commented on WooCommerce/WooCommerceUITests/Tests/OrdersTests.swift:
```diff
@@ -13,9 +14,18 @@ final class OrdersTests: XCTestCase {
+            let orders = try GetMocks.readOrdersData()
+            print(orders) // just for human validation
+
+            _ = try TabNavComponent()
```
Don't think we need to assign this to _ anymore. If it's still throwing an error then maybe a missing `@discardableResult` tag on one of the functions.

```suggestion
           try TabNavComponent()
```

thehenrybyrd replied:
Ah, yep! We needed `@discardableResult` on `verifyOrdersScreenLoaded`.

jostnes commented on WooCommerce/UITestsFoundation/Screens/Orders/OrdersScreen.swift:
```diff
@@ -37,4 +37,18 @@ public final class OrdersScreen: ScreenObject {
+        XCTAssertEqual(orders.count, app.tables.cells.count, "Expecting \(orders.count) orders, got \(app.tables.cells.count) instead!")
+        app.assertTextVisibilityCount(textToFind: String(orders[0].id))
+        app.assertTextVisibilityCount(textToFind: String(orders[0].total))
+        app.assertTwoTextsAppearOnSameLabel(firstSubstring: String(orders[0].id), secondSubstring: orders[0].billing.first_name)
```
This was [renamed/simplified recently](https://github.com/woocommerce/woocommerce-ios/blob/a1da511efd34f363cf2793d2766af77e8897df47/WooCommerce/UITestsFoundation/XCTest%2BExtensions.swift#L147) to `assertLabelContains`, I imagine this would fail after merging to trunk since it doesn't exist on that branch anymore.

thehenrybyrd replied:
Yep, it did! Sorted, thanks :)

jostnes reviewed (commented):
Nice work on the PR! 🎉 

Looks like there are hound issues to be fixed (found this doc on [swift lint](https://github.com/woocommerce/woocommerce-ios/blob/trunk/docs/coding-style-guide.md#swiftlint), maybe worth trying out locally. Though I personally haven't tried it, yet).  I've also added some small comments mostly around the use of comments in code and the leftover print statements.

pachlava commented on WooCommerce/WooCommerceUITests/Tests/OrdersTests.swift:
```diff
@@ -13,9 +14,18 @@ final class OrdersTests: XCTestCase {
+                .verifyOrdersScreenLoaded()
+                .verifyOrdersList(orders: orders) //parameter, variable
+                .selectOrder(atIndex: 0)//orders[0].id)
+                .verifySingleOrder(order: orders[0])
```
What happens here is selecting an order with index 0 in the list, and hoping that it will be the same as the order with index 0 in the mocks. While this will most probably be true forever, sorting is not what we test here, so I'd suggest opening the order by its name - by doing so we ensure that we are focused on testing that tapping a certain order opens this order, and do not rely on the assumption that the orders sorting in the mocks and the app is the same.

Also, wdyt about removing the comments at line 24 and 25?

pachlava commented on WooCommerce/WooCommerceUITests/Tests/OrdersTests.swift:
```diff
@@ -13,9 +14,18 @@ final class OrdersTests: XCTestCase {
+                .goBackToOrdersScreen()
+                .verifyOrdersScreenLoaded()
+        }
     }
```
```suggestion
    }
}
```

pachlava commented on WooCommerce/WooCommerceUITests/Flows/MockDataReader.swift:
```diff
@@ -51,4 +51,24 @@ class GetMocks {
+                    let formattedNumber = numberFormatter.string(from: NSNumber(value: Double(total)!))
+                    print("formatted total is: \(formattedNumber)")
+                    updatedData[index].total = formattedNumber!
+                }
```
It looks like the indentation of this block is two tabs deeper than it should be:

```suggestion
            let total = updatedData[index].total
            print("total is: \(total)")
            print("total after converted: \(Double(total) ?? 0)")
            let numberFormatter = NumberFormatter()
            numberFormatter.numberStyle = .decimal

            let formattedNumber = numberFormatter.string(from: NSNumber(value: Double(total)!))
            print("formatted total is: \(formattedNumber)")
            updatedData[index].total = formattedNumber!
        }
```

(you can also notice that I've changed the type for `total` to a constant, since it's never changed in the method)

thehenrybyrd replied:
Good call; I had initially made that `total` variable when I was figuring out how I'd account for formatting differences between mocks and the UI, but ended up not needing it.
Also fixed the indentation, thanks!

pachlava commented on WooCommerce/UITestsFoundation/Screens/Orders/SingleOrderScreen.swift:
```diff
@@ -8,11 +8,24 @@ public final class SingleOrderScreen: ScreenObject {
+    @discardableResult
+    public func verifySingleOrder(order: OrderData) throws -> Self {
+        app.assertTextVisibilityCount(textToFind: order.status, expectedCount: 1) //checks that the Mock order status appears on the screen
+        app.assertTextVisibilityCount(textToFind: order.billing.first_name, expectedCount: 2) // Check that the Mock first name ("Mira") appears twice - because this order has both billing and shipping details
```
In this case the explanation why the first name is expected to be shown twice might be helpful, so I'd keep it, but I'd make the comment shorter (we don't really need to mention the actual mock values to have the comment idea expressed):

```suggestion
        // Check that the first name appears twice - because the order screen has both billing and shipping details
        app.assertTextVisibilityCount(textToFind: order.billing.first_name, expectedCount: 2)
```

pachlava commented on WooCommerce/UITestsFoundation/Screens/Orders/SingleOrderScreen.swift:
```diff
@@ -8,11 +8,24 @@ public final class SingleOrderScreen: ScreenObject {
+    public func verifySingleOrder(order: OrderData) throws -> Self {
+        app.assertTextVisibilityCount(textToFind: order.status, expectedCount: 1) //checks that the Mock order status appears on the screen
+        app.assertTextVisibilityCount(textToFind: order.billing.first_name, expectedCount: 2) // Check that the Mock first name ("Mira") appears twice - because this order has both billing and shipping details
+        // try app.verifyChildElementOnParentCell(parent: "single-product-cell", child: "Mira") //this shouldn't work! "single-product-cell" doesn't include the text "Mira"
```
Generally, I find that once left, commented out or unused code has a tendency to stay there for a long time.

thehenrybyrd replied:
Yep, definitely need to clean those up. Thank you!

pachlava commented on WooCommerce/UITestsFoundation/Screens/Orders/SingleOrderScreen.swift:
```diff
@@ -8,11 +8,24 @@ public final class SingleOrderScreen: ScreenObject {
+        app.assertElementExistsOnCell(mainCell: "summary-table-view-cell", elementToFind: "\(order.billing.first_name) \(order.billing.last_name)")
+        //  "\(billing.first_name) + \(billing.last_name)"  "Mira Workman"
+        app.assertTextVisibilityCount(textToFind: order.total, expectedCount: 1)
+        XCTAssertTrue(app.staticTexts[order.line_items[0].name].isFullyVisibleOnScreen(), "First product name is not visible on screen!") //Black Coral shades is first in the mocks, that's what we're referencing, even though Malaya shades appear first on the screen.
```
What do you think of looping through `order.line_items` and checking that all products are there? This will improve test robustness, remove the need for the comments, and also should not be a big slowdown, as this requires no actions in UI.

jostnes replied:
Makes sense, I'll make the update so it loops through all products available on the order. I'm thinking of putting it in a for in loop that might not look so elegant so if there's a nicer way to do it, I'm all ears 😄

pachlava commented on WooCommerce/UITestsFoundation/Screens/Orders/SingleOrderScreen.swift:
```diff
@@ -8,11 +8,24 @@ public final class SingleOrderScreen: ScreenObject {
     }
 
+    @discardableResult
+    public func verifySingleOrder(order: OrderData) throws -> Self {
```
There's one check that this method lacks, and I think it's an important one - check for the order ID. Firstly, for consistency reasons (the order ID is checked in orders list). Secondly, why not, if we already have this data? And last: there's a (very low) probability that another order with the same client name and products will be opened instead of the expected one, and the check in its current form will pass.

pachlava reviewed (changes requested):
Hey @thehenrybyrd, thanks for working on this test! I executed it on iPhone 13, and it failed because the client first name was not shown twice in the single order screen, it was shown only once.

Apart from that, I left some review notes. Some of them I consider critical. I also noticed that you mention exact values from the mocks in the comments. The idea behind this setup: `have mocks` ~> `feed mocks values to Xcode test` ~> `Xcode checks that values it's been fed are displayed in the UI` implies that the tests can be isolated from the test data (with time, mocks can change, or we might drop WireMock in favour of something else). Because of this, I think the tests themselves should refrain from mentioning hardcoded data, and only serve as a mean of making sure that mocks data really make it to the UI.

thehenrybyrd committed:
Fixed lint issues

thehenrybyrd committed:
Remove excess `print` statements

They were to help me understand what was happening with the data while writing the test.

thehenrybyrd committed:
Update function names

A few functions had their names updated in other PRs and were no longer working.

thehenrybyrd committed:
Make `status` a constant in OrderDataStructs

Co-authored-by: Jos <jostnes@users.noreply.github.com>

thehenrybyrd committed:
Remove excess comments

And refine retained comments

thehenrybyrd committed:
Fixed indentation issues

thehenrybyrd committed:
Make `total` constant in MockDataReader

thehenrybyrd committed:
Remove unnecessary import

thehenrybyrd committed:
Remove `_ =` and add `@discardableResult` for `verifyOrdersScreenLoaded`

thehenrybyrd commented:
@pachlava:

> I executed it on iPhone 13, and it failed because the client first name was not shown twice in the single order screen, it was shown only once.

Thanks for pointing this out! I'm seeing this now as well, and it's a valid failure. This is a new bug, likely related to work being done on Order Creation. I've created an issue at https://github.com/woocommerce/woocommerce-ios/issues/5894 for that.

The test also doesn't run at the moment due to the new "Menu" tab - part of the [Home Screen Updates Project](https://github.com/woocommerce/woocommerce-ios/projects/43).

For now, the test will run if you:
1. Comment out all references to and functions for reviews in `TabNavComponent` and `ReviewsTests`
2. Change `expectedCount` to `1` in `SingleOrderScreen` line 19 (`assertTextVisibilityCount` for `order.billing.first_name`).

thehenrybyrd commented:
@pachlava:

> The idea behind this setup: `have mocks` ~> `feed mocks values to Xcode test` ~> `Xcode checks that values it's been fed are displayed in the UI` implies that the tests can be isolated from the test data... I think the tests themselves should refrain from mentioning hardcoded data, and only serve as a mean of making sure that mocks data really make it to the UI.

That makes a lot of sense to me! I've tried to reference the mock data whenever I can, but please point out if there is still hardcoded data I've missed after cleaning this up.

I also really like your suggestions for referencing the singleOrder by name, looping through the line_items, and checking the order ID for the singleOrder. Hope to be able to implement those! @jostnes may help out with that if she's got time, otherwise I'll work on those when I'm back from AFK.

jostnes committed:
select by orderNumber, loop all products, check orderNumber

jostnes committed:
Merge branch 'trunk' into add/orders-ui-tests

jostnes committed:
fix indentation, update for loop condition

jostnes committed:
Merge branch 'trunk' into add/orders-ui-tests

jostnes committed:
add missing func

jostnes committed:
remove whitespace

pachlava commented on WooCommerce/WooCommerceUITests/Tests/OrdersTests.swift:
```diff
@@ -11,11 +11,20 @@ final class OrdersTests: XCTestCase {
         app.launch()
         try LoginFlow.logInWithWPcom()
+
+        try TabNavComponent()
+            .goToOrdersScreen()
```
I'd vote for moving this inside the test - this will save us one screen instantiation (~9 sec in this case). Currently, `OrdersScreen` is instantiated twice in a row.

jostnes replied:
Ah right, didn't know about the consequence of moving it here, the idea behind it is in the case where the Orders tests expand we don't have to add to each test as it is part of the `setup` already. But if it adds ~9 seconds then moving it into the test would be better.

pachlava commented on WooCommerce/UITestsFoundation/Screens/Orders/SingleOrderScreen.swift:
```diff
@@ -8,11 +8,37 @@ public final class SingleOrderScreen: ScreenObject {
+        app.assertTextVisibilityCount(textToFind: order.status, expectedCount: 1)
+        app.assertTextVisibilityCount(textToFind: order.total, expectedCount: 1)
+
+        // Adding # to textToFind to eliminate the case where number matches product price
```
I don't think this comment is necessary - it's visible on the screen that orders numbers are preceded with `#`, so adding `#` would make total sense, IMO, even without a possible case of coincidence with price.

```suggestion
```

pachlava commented on WooCommerce/UITestsFoundation/Screens/Orders/SingleOrderScreen.swift:
```diff
@@ -8,11 +8,37 @@ public final class SingleOrderScreen: ScreenObject {
+        // Expects 2 instances of order.number - one in Header and one in Summary
+        app.assertTextVisibilityCount(textToFind: "#\(order.number)", expectedCount: 2)
+
+        // Temporary excluding until GH Issue is fixed: https://github.com/woocommerce/woocommerce-ios/issues/5894
+        let issueFixed = false
+        if issueFixed == true {
+            // Expects 2 instances of first_name - one in Summary and one in Shipping details
+            app.assertTextVisibilityCount(textToFind: order.billing.first_name, expectedCount: 2)
+        }
```
An interesting point is that currently the first name was shown twice for me, once I enabled this check.

jostnes replied:
Interesting, I'll retest and remove this condition if that's no longer an issue. Thanks for taking the time to check both conditions @pachlava !

pachlava commented on WooCommerce/UITestsFoundation/Screens/Orders/SingleOrderScreen.swift:
```diff
@@ -8,11 +8,37 @@ public final class SingleOrderScreen: ScreenObject {
+
+        app.assertElement(matching: "summary-table-view-cell", existsOnCellWithIdentifier: "\(order.billing.first_name) \(order.billing.last_name)")
+
+        // Loops through all products on the order
+        for index in 0..<order.line_items.count {
+            XCTAssertTrue(app.staticTexts[order.line_items[index].name].isFullyVisibleOnScreen(), "'\(order.line_items[index].name)' is missing!")
+        }
```
This can be shortened a bit, and with more specific naming the comment becomes redundant:

```suggestion
        for product in order.line_items {
            XCTAssertTrue(app.staticTexts[product.name].isFullyVisibleOnScreen(), "'\(product.name)' is missing!")
        }
```

jostnes replied:
I think it also becomes easier to read this way, thanks for the suggestion!

pachlava reviewed (approved):
Leaving a few speed/formatting nitpicks, but good to go in overall. Thanks for taking this task @jostnes !

jostnes committed:
simplify and move to get speed improvements
